resources/feature_dism.rb

```diff
@@ -18,87 +18,157 @@
-    @installed ||= begin
-      install_ohai_plugin unless node['dism_features']
+  # @return [Array] features the user has requested to install which need installation
+  def features_to_install
+    # the intersection of the features to install & disabled features are what needs installing
+    @install ||= new_resource.feature_name & node['dism_features_cache']['disabled']
+  end
+
+  # @return [Array] features the user has requested to remove which need removing
+  def features_to_remove
+    # the intersection of the features to remove & enabled features are what needs removing
+    @remove ||= new_resource.feature_name & node['dism_features_cache']['enabled']
+  end
 
-      # Compare against ohai plugin instead of costly dism run
-      node['dism_features'].key?(new_resource.feature_name) && node['dism_features'][new_resource.feature_name] =~ /Enable/
+  # @return [Array] features the user has requested to delete which need deleting
+  def features_to_delete
+    # the intersection of the features to remove & enabled/disabled features are what needs removing
+    @remove ||= begin
+      all_available = node['dism_features_cache']['enabled'] +
+                      node['dism_features_cache']['disabled']
+      new_resource.feature_name & all_available
     end
   end
 
-  def available?
-    @available ||= begin
-      install_ohai_plugin unless node['dism_features']
+  # if any features are not supported on this release of Windows or
+  # have been deleted raise with a friendly message. At one point in time
+  # we just warned, but this goes against the behavior of ever other package
+  # provider in Chef and it isn't clear what you'd want if you passed an array
+  # and some features were available and others were not.
+  # @return [void]
+  def fail_if_unavailable
+    all_available = node['dism_features_cache']['enabled'] +
+                    node['dism_features_cache']['disabled'] +
+                    node['dism_features_cache']['removed']
 
-      # Compare against ohai plugin instead of costly dism run
-      node['dism_features'].key?(new_resource.feature_name) && node['dism_features'][new_resource.feature_name] !~ /with payload removed/
-    end
+    # the difference of desired features to install to all features is what's not available
+    unavailable = (new_resource.feature_name - all_available)
+    raise "The Windows feature#{'s' if unavailable.count > 1} #{unavailable.join(',')} #{unavailable.count > 1 ? 'are' : 'is'} not available on this version of Windows. Run 'dism /online /Get-Features' to see the list of available feature names." unless unavailable.empty?
   end
 
-  def reload_ohai_features_plugin(take_action, feature_name)
-    ohai "Reloading Dism_Features Plugin - Action #{take_action} of feature #{feature_name}" do
-      action :reload
-      plugin 'dism_features'
+  # run dism.exe to get a list of all available features and their state
+  # and save that to the node at node.normal (same as ohai) level.
```

Comment:
Hello @tas50 

Isn't ohai using `automatic` level? Why did you chose `normal`?
Would you accept a PR to pass it to something non-persistent (`automatic` of `force_override`)?

Nice feature btw.

Reply:
Oh man. Giant facepalm here. That should be automatic. I'll get a fix for that in ASAP

Reply:
No hurry on my side, I don't plan to upgrade the *windows* soon, but I agree this should be automatic.
Thanks for the quick answer.

Reply:
Looks like you can't set automatic from within the cookbook actually. I'm just going to make these default since there's no threat of a namespace collision and it would potentially allow someone to do some magical things there is they wished to (but probably shouldn't)

Reply:
Scratch that. Override it is so we can override a normal attribute set by 4.0.0

Reply:
Ok so I saw you released the new version with the `override` level.
But just out of curiosity why are you saying `you can't set automatic from within the cookbook`?
I'm doing it a lot!

See https://github.com/Annih/chef-pci/blob/master/attributes/automatic.rb#L6 for instance

Reply:
At least on 12.7 it was just throwing an empty hash. I'll take a peak at exactly what was going on there once I wrap up some other work.